### PR TITLE
Doc - Explain how to extend ngeo

### PR DESCRIPTION
**Pull request by ger-benjamin**
As asked.
Should be the same for 2.5, 2.6

### CODE REVIEW

File: doc/integrator/ngeo.rst.mako [hunk 1]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+Create your own component
+-------------------------
+
+Create your ``.js`` and ``.html`` (partial) files in a dedicated folder under:
+``geoportal/<package>_geoportal/static-ngeo/js/``. We encourage you to write standard ECMAScript and
+use AngularJS only to display your component.
+
+For the structure, you can be inspired by one of the components in ngeo.
+
+You should register each your AngularJS element (component, controller, service, etc.) in a module. First
```
Comment (asaunier) [inline doc/integrator/ngeo.rst.mako:126]:
```suggestion
You should register each of your AngularJS element (component, controller, service, etc.) in a module. First
```

File: doc/integrator/ngeo.rst.mako [hunk 2]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
 You also have a ``default`` entry from where we can give the base values for ``constants``,
 ``dynamic_constants``, ``static`` and ``routes``. These values will be updated with the interface configuration
 (but can not be removed via interface configuration).
+
+Create your own component
```
Comment (asaunier) [inline doc/integrator/ngeo.rst.mako:117]:
```suggestion
Creating your own component
```

File: doc/integrator/ngeo.rst.mako [hunk 3]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+If you have no error and no result, verify that the module is correctly declared and loaded. AngularJS can't
+throw errors for a component that it doesn't know and having a custom markup in the html is not an error
+for the html.
+
+.. note::
+
+To load a partial (html) file for a component you can use this syntaxe inside the declaration of your
+component: ``template: require('./component.html')`` (from ``./Component.js`` file).
+
+Extend or override an existing component
```
Comment (asaunier) [inline doc/integrator/ngeo.rst.mako:177]:
```suggestion
Extending or overridding an existing component
```

File: doc/integrator/ngeo.rst.mako [hunk 4]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+----------------------------------------
+
+Overriding an existing partial (html) or js element is possible but you have to test carefully your code
```
Comment (asaunier) [inline doc/integrator/ngeo.rst.mako:180]:
```suggestion
Overridding an existing partial (html) or js element is possible but you have to test carefully your code
```

File: doc/integrator/ngeo.rst.mako [hunk 5]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+   DrawController.prototype.newFunction = function () {};
+
+To replace an existing function, just create a function with the same name that the one to replace:
+
+.. code:: javascript
+
+   DrawController.prototype.$onInit = function () {};
+
+You will have to copy-paste the content of the original function to keep the component working. Then you
+can extend it.
```
Comment (asaunier) [inline doc/integrator/ngeo.rst.mako:203]: Is it possible to override the original function by calling it in the new function, without copy-pasting the whole function?

**Review by asaunier — COMMENTED** [review]
Can't say about the actual coding, my comments are only about the text itself + question.
Looks good at first sight, thanks :)

File: doc/integrator/ngeo.rst.mako (hunk 5 above)
Comment (ger-benjamin) [reply doc/integrator/ngeo.rst.mako:203]: No, you attach a whole new function at the place of the old one. The old one is lost.

You can use Inheritance for that, but that means you have to rewrite a constructor (some lines) and replace the "normal" controller with your new "children" controller in ngeo (one line).
So.. well... it's not a bad idea. But I've not tested.

Should I add a sentence that says it's possible ?

Comment (ger-benjamin) [reply doc/integrator/ngeo.rst.mako:203]: Oh, and you can't do that if you want to interfere between lines of the functions to override.

File: doc/integrator/ngeo.rst.mako (hunk 1 above)
Comment (ger-benjamin) [reply doc/integrator/ngeo.rst.mako]: Done

File: doc/integrator/ngeo.rst.mako (hunk 2 above)
Comment (ger-benjamin) [reply doc/integrator/ngeo.rst.mako]: Done

File: doc/integrator/ngeo.rst.mako (hunk 3 above)
Comment (ger-benjamin) [reply doc/integrator/ngeo.rst.mako]: Done

File: doc/integrator/ngeo.rst.mako (hunk 4 above)
Comment (ger-benjamin) [reply doc/integrator/ngeo.rst.mako]: Done

**Comment by ger-benjamin** [comment]
Thanks @asaunier

File: doc/integrator/ngeo.rst.mako (hunk 5 above)
Comment (asaunier) [reply doc/integrator/ngeo.rst.mako:203]: OK, no forget it, thanks

File: doc/integrator/ngeo.rst.mako [hunk 6]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+
+   <project-directivename></project-directivename>
+
+If you have no error and no result, verify that the module is correctly declared and loaded. AngularJS can't
+throw errors for a component that it doesn't know and having a custom markup in the HTML is not an error
+for the HTML.
+
+.. note::
+
+   To load a partial (HTML) file for a component you can use this syntaxe inside the declaration of your
```
Comment (jwkaltz) [inline doc/integrator/ngeo.rst.mako:174]: syntaxe -> syntax

File: doc/integrator/ngeo.rst.mako [hunk 7]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+   component: ``template: require('./component.html')`` (from ``./Component.js`` file).
+
+Extending or overridding an existing component
+----------------------------------------------
+
+Overridding an existing partial (HTML) or JS element is possible but you have to test carefully your code
```
Comment (jwkaltz) [inline doc/integrator/ngeo.rst.mako:180]: Overridding -> Overriding

File: doc/integrator/ngeo.rst.mako [hunk 8]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+    
+   import `import {DrawController} from 'ngeo/draw/Controller.js'`
+   /**
+    * Doctype is important for compilation.
+    * @type {string}
+    */
+   DrawController.prototype.myNewValue;
+   DrawController.prototype.newFunction = function () {};
+
+To replace an existing function, just create a function with the same name that the one to replace:
```
Comment (jwkaltz) [inline doc/integrator/ngeo.rst.mako:196]: that the -> than the

File: doc/integrator/ngeo.rst.mako [hunk 9]
```diff
@@ -113,3 +113,112 @@ The dynamic values names are: ``interface``, ``cache_version``, ``lang_urls``, `
...
+
+.. code:: javascript
+
+   DrawController.prototype.$onInit = function () {};
+
+You will have to copy-paste the content of the original function to keep the component working. Then you
+can extend it.
+
+To override a partial (HMTL) file, you have to copy-paste the whole original file. Then you can adapt it.
+To make the original component using your own partial file, you must set the template cache of the module:
```
Comment (jwkaltz) [inline doc/integrator/ngeo.rst.mako:206]: using -> use

**Review by jwkaltz — APPROVED** [review]

**Pull request merged**